### PR TITLE
Change list striding parameters to take optional integer (preceding work for #49352)

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49352 Striding for lists Part 2
* **#48719 Striding for lists Part 1**

Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25929902](https://our.internmc.facebook.com/intern/diff/D25929902)

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
What is actually changing here? It would be nice to have the codemod in a separate PR.

**Comment by eellison** [comment]
Yea, agreed with alban, it might be easier to review this once the whole stack is out

**Comment by eellison** [comment]
cc @ailzhang XLA failure

**Comment by eellison** [comment]
Some context: we are fixing the lists slicing logic with a negative index. Part of this change involves changing the ir emission for aten::slice to emit None for non-present step value instead of INT_MAX. Rather than have special case logic for all of the builtins type that may or may not support negative striding (lists, tensors, strs) we are updating their implementations so that they can all take in optionals. Actual negative slicing of Tensors is still NYI, but making the schema changes here makes future implementation easier and makes the logic in ir emission cleaner.

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 1]
```diff
@@ -340,7 +340,12 @@ Tensor select_backward_batching_rule(const Tensor& grad, IntArrayRef input_sizes
   return grad_physical.newLogicalFromPhysical(grad_input);
 }
 
-Tensor slice_batching_rule(const Tensor& self, int64_t dim, int64_t start, int64_t end, int64_t step) {
+Tensor slice_batching_rule(
```
Comment (eellison) [inline aten/src/ATen/BatchingRegistrations.cpp:343]: cc: @zou3519 , i think these changes are harmless but double checking

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -1070,36 +1070,52 @@ Tensor index_select_sparse(const Tensor& self, int64_t dim, const Tensor& index)
   }
 }
 
-Tensor slice(const Tensor& self, int64_t dim, int64_t start, int64_t end, int64_t step) {
+Tensor slice(
+    const Tensor& self,
+    int64_t dim,
+    c10::optional<int64_t> start,
+    c10::optional<int64_t> end,
+    c10::optional<int64_t> step) {
```
Comment (eellison) [inline aten/src/ATen/native/TensorShape.cpp:1078]: In python, `[::]` corresponds to None, None, 1 right ? I don't think we actually need to make step an optional

File: aten/src/ATen/native/TensorShape.cpp [hunk 3]
```diff
@@ -1070,36 +1070,52 @@ Tensor index_select_sparse(const Tensor& self, int64_t dim, const Tensor& index)
...
   auto sizes = self.sizes().vec();
   auto strides = self.strides().vec();
+
+  // handle optinal parameters
+  int64_t start_val = start.has_value() ? start.value() : 0;
+  int64_t end_val = end.has_value() ? end.value() : INT64_MAX;
+  int64_t step_val = step.has_value() ? step.value() : 1;
+
   // TODO: support negative strides
-  TORCH_CHECK(step > 0, "slice step must be positive");
```
Comment (eellison) [inline aten/src/ATen/native/TensorShape.cpp:1082]: Why were these lines changed ? I think we should leave negative indexing NYI for tensors, it might be tricky, and i think is out of scope for this PR. Just error'ing shoudl be fine.

**Review by eellison — COMMENTED** [review]
Looks great! Nice job wading through the tracer code.

I have two comments:
- I don't think we actually need to make `step` optional, I think python just defaults it to 1
- I'm not sure about the `TenssorShape.cpp` changes, I think it would be better to error at negative step within TensorShape instead of trying to implement it for this PR.

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:343]: Yup, this lgtm

File: aten/src/ATen/native/TensorShape.cpp (hunk 3 above)
Comment (tugsbayasgalan) [reply aten/src/ATen/native/TensorShape.cpp:1082]: I think it was just clang formatting

Comment (eellison) [reply aten/src/ATen/native/TensorShape.cpp:1082]: oh, nvm then my bad. Changes heree look good 👍

File: aten/src/ATen/native/TensorShape.cpp [hunk 4]
```diff
@@ -1070,36 +1070,52 @@ Tensor index_select_sparse(const Tensor& self, int64_t dim, const Tensor& index)
...
+    c10::optional<int64_t> step) {
   int64_t ndim = self.dim();
   if (ndim == 0) {
     TORCH_CHECK_INDEX(false, "slice() cannot be applied to a 0-dim tensor.");
   }
   dim = maybe_wrap_dim(dim, ndim);
   auto sizes = self.sizes().vec();
   auto strides = self.strides().vec();
+
+  // handle optinal parameters
```
Comment (eellison) [inline aten/src/ATen/native/TensorShape.cpp:1087]: nit: /s/optinal/optional

Comment (eellison) [reply aten/src/ATen/native/TensorShape.cpp:1120]: bump lol

**Review by eellison — APPROVED** [review]
LGTM, Ship it! Great job! We might want to sanity check the same forward-compatibility checks we did previously with the newest changes, just to be safe. But up to you.

File: torch/csrc/jit/runtime/register_prim_ops.cpp [hunk 5]
```diff
@@ -29,23 +30,25 @@ namespace {
 
 std::string stringSlice(
```
Comment (eellison) [inline torch/csrc/jit/runtime/register_prim_ops.cpp:31]: Actually, sorry, one more comment: just as we updated the list implementation to work with negative step we should do the same with strings right ?

**Review by eellison — APPROVED** [review]
LGTM, ship it! We should wait a couple weeks or so before landing PT2 so we dont run into FC issues.

File: torch/csrc/jit/runtime/register_prim_ops.cpp [hunk 6]
```diff
@@ -29,23 +31,22 @@ namespace {
...
     std::string string,
-    int64_t start,
-    int64_t end,
+    c10::optional<int64_t> start,
+    c10::optional<int64_t> end,
     int64_t step) {
-  TORCH_CHECK(step == 1, "Slicing a string only supports step=1");
-
-  const int64_t size = string.size();
+  int64_t start_val = start.has_value() ? start.value() : INT64_MAX;
```
Comment (ezyang) [inline torch/csrc/jit/runtime/register_prim_ops.cpp:37]: Why does this default to `INT64_MAX`? Why not just set it to 0? cc @eellison this is impeding SymInt-ification

Comment (eellison) [reply torch/csrc/jit/runtime/register_prim_ops.cpp:37]: Ugh, I forget, but I think that was mirroring how the python slice handling code went where `None` would be encoded as `INT64_MAX`
